# 03_reddit_analysis
Having thoroughly analysed the data obtained from Twitter, we now move on to Reddit. We are keen to make some strong comparisons to our Twitter analysis. This is why we will try and conduct this reddit data investigation with identical, or at least similar units of analysis to our Twitter report. This means, we will...

- run sentiment analysis on text in reddit posts (not in the links)
- do some wordclouds for negative and positive words. also frequency tables for these words
- we will look at the URLs contained in our reddit posts. 
    - what's the proportion going to reddit vs. outside of reddit?
    - frequency table of most frequent urls / domains. we may have to expand a bunch of urls
    - finally, compare the urls and domains we saw a lot on twitter. is there going to be some significant overlap with the urls we saw on twitter? 

Overall, we collected top-level posts (i.e. no comments) for 5 distinct subreddits, which we considered as relevant to the world cup a priori: 
`FIFAWorldCupQATAR22, football, qatar, soccer, worldcup`

NL, 12/01/23

### IMPORTS

In [80]:
import os
import json
import pandas as pd
from dateutil import parser as date_parser
import re
from urllib.parse import urlparse

### PATHS & CONSTANTS

In [3]:
SUBREDDITS = ['FIFAWorldCupQATAR22','football','qatar','soccer','worldcup']
REDDIT_PATH = '/home/nikloynes/projects/world_cup_misinfo_tracking/data/reddit_posts/'
EXPORT_PATH = '/home/nikloynes/projects/world_cup_misinfo_tracking/data/exports/reddit/'

### INIT

In [75]:
pd.set_option('display.max_rows', 100)
pd.set_option('display.max_columns', 50)

### FUNCTIONS

In [83]:
def extract_urls(text:str) -> list:
    '''
    extracts urls from text
    returns all urls in text in a list
    '''
    urls = re.findall("http[s]?://(?:[a-zA-Z]|[0-9]|[$-_@.&+]|[!*\(\),]|(?:%[0-9a-fA-F][0-9a-fA-F]))+", text)

    return urls


### THE THING!

Let's list all our reddit files

In [10]:
reddit_files = []

for r in SUBREDDITS:
    reddit_files += [REDDIT_PATH+r+'/'+x for x in os.listdir(REDDIT_PATH+r+'/')]

We have data collected in hourly chunks for each of our 5 subreddits.  
This leaves us with a total of 4096 files. Let's quickly check how many reddit posts that represents.

In [12]:
counter = 0
for file in reddit_files:
    with open(file, 'r') as infile:
        for line in infile:
            counter += 1

In [13]:
counter

18711

Great, so we have collected data for a total of `18,711` reddit posts. Clearly, this is quite a bit less than we got for Twitter, but we did follow a completely different strategy: rather than filtering the entirety of Reddit for a bunch of keywords, we instead tracked 5 subreddits. This is obviously also an artefact of how reddit's API works, and indeed how the platform works.

OK, let's just refresh our memories -- what does a given reddit post look like in json format? 

In [20]:
for file in reddit_files:
    with open(file, 'r') as infile:
        for line in infile:
            tmp = json.loads(line)
            break
    break

In [15]:
tmp

{'id': 'yyt2cj',
 'created_utc': '2022/11/18 19:58:45',
 'title': 'Watch FIFA World Cup QATAR 2022 Live Stream Online Without Cable.',
 'selftext': "Welcome to **Watch [FIFA World Cup Qatar 2022 Live](https://fifa-world-cup-2022-live-streams.blogspot.com/) Stream Online for Free HD TV** coverage match online from here. Everybody knows that (FIFA World Cup) games means excitation. Depending on the country you in, find available options to watch FIFA World Cup games online. So, you can easily watch FIFA World Cup Qatar 2022 live matches from here.\n\n\n#Watch : [FIFA World Cup Qatar 2022 Live Stream Free](https://fifa-world-cup-2022-live-streams.blogspot.com/)\n\n\nYou can easily watch your favorite teams FIFA World Cup live streaming online free. This is a free FIFA World Cup streaming website that's provide multiple links to watch any FIFA World Cup game live. We offer the best FIFA World Cup live streams link in HD/HQ/4K. Watch the FIFA World Cup all game from your mobile, tablet, Mac

First off, it's quite funny that our very first reddit post by chance seems to be telling the same story we observed via our tweet collections. Ignoring that, however, this gives us a good idea of what we want to aggregate over the entirety of our available reddit data.

In [57]:
on_reddit_domains = [
    'i.redd.it',
    'self.worldcup',
    'self.qatar',
    'v.redd.it',
    'self.football',
    'self.FIFAWorldCupQATAR22',
    'self.soccer',
    'reddit.com',
]

In [87]:
urls_df = pd.DataFrame()
domains_freq = {}
user_overview = []
unique_ids = [] # we've noticed that we've collected some duplicate data. not really a problem, but let's make sure we don't double-count
urls_in_text_df = pd.DataFrame()

for file in reddit_files:
    with open(file, 'r') as infile:
        for line in infile:
            tmp = json.loads(line)
            
            if tmp['id'] not in unique_ids:
                unique_ids.append(tmp['id'])
            else:
                # we have already seen this id. let's move on
                continue
              
            # urls
            internal = False
            if tmp['domain'] in on_reddit_domains:
                internal = True
            
            tmp_url = {
                'url' : tmp['url'],
                'domain' : tmp['domain'],
                'internal_link' : internal,
                'timestamp' : date_parser.parse(tmp['created_utc']),
                'user_name' : tmp['user_name'],
                'user_id' : tmp['user_id']                 
            }
            urls_df = pd.concat([urls_df, pd.DataFrame([tmp_url])])

            if tmp['domain'] not in domains_freq.keys():
                domains_freq[tmp['domain']] = 1
            else:
                domains_freq[tmp['domain']] += 1

            # we are also interested in URLs contained in the 
            # 'selftext' field in our reddit post object-- as our
            # example from above contains some which were not captured
            # in the object itself.
            embedded_urls = extract_urls(tmp['selftext'])
            for eurl in embedded_urls:
                try:
                    tmp_domain = urlparse(eurl).netloc
                except ValueError:
                    tmp_domain = None
                tmp_eurl = {
                    'url' : eurl,
                    'domain' : tmp_domain,
                    'timestamp' : date_parser.parse(tmp['created_utc']),
                    'user_name' : tmp['user_name'],
                    'user_id' : tmp['user_id']
                }
                urls_in_text_df = pd.concat([urls_in_text_df, pd.DataFrame([tmp_eurl])]) 
            
            if tmp['user_id'] not in [x['user_id'] for x in user_overview]:
                if tmp['user_created_utc']!=None:
                    user_created = date_parser.parse(tmp['user_created_utc'])
                else:
                    user_created = None
                tmp_user = {
                    'user_id' : tmp['user_id'],
                    'user_name' : tmp['user_name'],
                    'user_created_at' : user_created,
                    'user_comments_total' : tmp['num_comments'],
                    'user_karma_total' : tmp['user_total_karma'],
                    'user_upvotes_total' : tmp['ups'],
                    'user_downvotes_total' : tmp['downs'],
                    'total_wc_posts' : 1
                }
                user_overview.append(tmp_user)
            else:
                # first find the index where this is the case
                the_index = next((index for (index, d) in enumerate(user_overview) if d['user_id'] == tmp['user_id']), None)

                # then make changes in this entry
                user_overview[the_index]['user_comments_total'] += tmp['num_comments']
                user_overview[the_index]['user_upvotes_total'] += tmp['ups']
                user_overview[the_index]['user_downvotes_total'] += tmp['downs']
                user_overview[the_index]['total_wc_posts'] += 1
                
                # the karma recorded in our jsons is already cumulative. so it could only be
                # that a user's karma from a previous json is outdated, so we need to check if
                # it's grown and only then update this value. 
                try:
                    if tmp['user_total_karma'] > user_overview[the_index]['user_karma_total']:
                        user_overview[the_index]['user_karma_total'] = tmp['user_total_karma']
                except TypeError as e:
                    print(f'Error encountered when comparing user karma. Skipping.')
                    print(e)

Error encountered when comparing user karma. Skipping.
'>' not supported between instances of 'NoneType' and 'int'
Error encountered when comparing user karma. Skipping.
'>' not supported between instances of 'NoneType' and 'int'
Error encountered when comparing user karma. Skipping.
'>' not supported between instances of 'NoneType' and 'int'
Error encountered when comparing user karma. Skipping.
'>' not supported between instances of 'NoneType' and 'int'
Error encountered when comparing user karma. Skipping.
'>' not supported between instances of 'NoneType' and 'int'
Error encountered when comparing user karma. Skipping.
'>' not supported between instances of 'NoneType' and 'int'
Error encountered when comparing user karma. Skipping.
'>' not supported between instances of 'NoneType' and 'int'
Error encountered when comparing user karma. Skipping.
'>' not supported between instances of 'NoneType' and 'int'
Error encountered when comparing user karma. Skipping.
'>' not supported between

In [88]:
# now some formatting/cleaning on the data we've just built
domains_freq = dict(sorted(domains_freq.items(), key=lambda x:x[1], reverse=True))
domains_freq_df = pd.DataFrame([domains_freq]).transpose().reset_index().rename(columns={'index' : 'domain', 0 : 'freq'})
domains_freq_df.loc[domains_freq_df['domain'].isin(on_reddit_domains), 'internal'] = True
domains_freq_df['internal'].fillna(False, inplace=True)

user_overview_df = pd.DataFrame(user_overview)

Let's now have a look at the dataframe we've created.

We've recorded `15,049` URL sharing incidences.

In [89]:
urls_df['internal_link'].value_counts()

True     9563
False    5486
Name: internal_link, dtype: int64

In [90]:
urls_df['internal_link'].value_counts(normalize=True)

True     0.635458
False    0.364542
Name: internal_link, dtype: float64

Out of these URL shares, `64%` (`9563 shares`) were to internal Reddit resources, while `36%` (`5486 shares`) were to external URLs.

Let's have a look at our unique domains

In [91]:
domains_freq_df[:100]

domain  freq  internal
0                i.redd.it  2555      True
1            self.worldcup  2185      True
2                v.redd.it  1430      True
3               self.qatar  1400      True
4              twitter.com  1123     False
5            self.football  1083      True
6              self.soccer   591      True
7             streamja.com   513     False
8              streamin.me   467     False
9                 youtu.be   321     False
10              reddit.com   318      True
11             youtube.com   273     False
12                 dubz.co   264     False
13             i.imgur.com    89     False
14         theguardian.com    80     False
15          streamable.com    69     False
16                fifa.com    64     False
17               bbc.co.uk    64     False
18               marca.com    62     False
19            streamff.com    62     False
20         dailymail.co.uk    56     False
21         theathletic.com    47     False
22            streamag.com    37     False
23                 bbc.com    35     False
24            townflex.com    33     False
25         telegraph.co.uk    32     False
26                espn.com    29     False
27             reuters.com    28     False
28             bolavip.com    28     False
29              apnews.com    27     False
30                goal.com    26     False
31            ge.globo.com    26     False
32            mirror.co.uk    24     False
33              lequipe.fr    24     False
34     football-italia.net    24     False
35               imgur.com    23     False
36          thetimes.co.uk    21     False
37  saturdays.substack.com    19     False
38                 cnn.com    17     False
39             lhknews.com    17     False
40           skysports.com    16     False
41    gianlucadimarzio.com    15     False
42              relevo.com    15     False
43      useyourtalent.info    13     False
44                  hln.be    13     False
45             nytimes.com    12     False
46       voetbalprimeur.nl    12     False
47        13thmansports.ca    11     False
48               record.pt    11     False
49               vavel.com    11     False
50                abola.pt    11     False
51          hellogists.com    11     False
52           aljazeera.com    10     False
53           instagram.com    10     False
54               kicker.de    10     False
55                 bild.de    10     False
56                  as.com    10     False
57           tycsports.com    10     False
58                   vi.nl    10     False
59              tiktok.com     9     False
60        en.wikipedia.org     9     False
61  streamingdigitally.com     9     False
62           talksport.com     8     False
63       independent.co.uk     8     False
64      bleacherreport.com     8     False
65            thesun.co.uk     8     False
66                  ft.com     8     False
67             gazzetta.it     8     False
68            telegraaf.nl     8     False
69        footgameblog.com     8     False
70          viralnewsy.com     8     False
71           express.co.uk     7     False
72      washingtonpost.com     7     False
73           vm.tiktok.com     7     False
74       bunsheesports.com     7     False
75   self.IBetYouCommunity     7     False
76                 1908.nl     7     False
77       streamscores.link     7     False
78                  nos.nl     7     False
79         link.medium.com     6     False
80            news.sky.com     6     False
81        sportingnews.com     6     False
82               90min.com     6     False
83              google.com     6     False
84              uol.com.br     6     False
85                ebay.com     6     False
86         24hfootnews.com     6     False
87            pixstory.com     6     False
88      theathleteshub.org     5     False
89              nypost.com     5     False
90     sportsgazette.co.uk     5     False
91                 msn.com     5     False
92     uk.sports.yaho

Interesting. We again see quite a lot of apparent links to streaming sites... albeit there appear to be fewer of them compared to what we saw on Twitter. Let's dig into this a bit more. 

Let's now have a look at URLs contained within the `selftext` field of a given reddit post.

In [104]:
urls_in_text_freq_df = urls_in_text_df.groupby('domain').count().sort_values('url', ascending=False).reset_index()[['domain', 'url']].rename(columns={'url' : 'freq'})

In [107]:
len(urls_in_text_df)

2773

In [108]:
len(urls_in_text_freq_df)

175

OK, so when regarding only URLs contained directly in text, we have a further `2,773` URL posting incidences; resulting in `175` unique domains from these URL shares.

Let's have a look at the most popular of those shared domains

In [109]:
urls_in_text_freq_df[:100]

domain  freq
0                                  www.reddit.com   767
1                                    www.espn.com   301
2                                      reddit.com   299
3                           www.reddit-stream.com   295
4                                 preview.redd.it   239
5                                          k2s.cc   130
6                                      discord.gg   109
7                                  old.reddit.com    91
8                                  new.reddit.com    88
9                                 www.youtube.com    71
10                                       youtu.be    45
11                                      imgur.com    28
12                                    twitter.com    20
13                               en.wikipedia.org    14
14                                    i.imgur.com    10
15                              reddit-stream.com    10
16                            www.timeanddate.com    10
17                                docs.google.com     8
18                           www.livesoccertv.com     6
19                                 www.google.com     5
20                            www.theguardian.com     5
21                                   www.ncaa.com     4
22                                theathletic.com     4
23                 umassamherst.co1.qualtrics.com     4
24                          www.independent.co.uk     4
25                            www.dailymail.co.uk     4
26                                www.nytimes.com     4
27                                    youtube.com     4
28                            digitalhub.fifa.com     4
29                                link.medium.com     3
30                         www.prodirectsport.com     3
31                           hospitality.fifa.com     3
32                                   www.fifa.com     3
33                                      i.redd.it     3
34                                    streamin.me     3
35                               livesoccertv.com     3
36                        stats.statbroadcast.com     2
37                                    www.bbc.com     2
38                                  www.bbc.co.uk     2
39                                www.infobae.com     2
40                      soccerscience.gumroad.com     2
41                              www.instagram.com     2
42                               KnowRivalry.com)     2
43                              www.eurosport.com     2
44                                  strawpoll.com     2
45                                 visitqatar.com     2
46                                  www.90min.com     2
47                             thegodofsports.com     2
48                                    tribuna.com     2
49                                  candorium.com     2
50                                        chng.it     2
51                                    uk.puma.com     2
52                              bunsheesports.com     2
53                                  play.fifa.com     2
54                                   www.goal.com     2
55                                newscraving.com     2
56                                      forms.gle     2
57                             www.elfutbolero.us     2
58                                     www.dfb.de     2
59                            www.dailymotion.com     2
60                      www.mercurialfootball.com     2
61  fifa-world-cup-2022-live-streams.blogspot.com     2
62                             en.m.wikipedia.org     2
63                          www.redditraffler.com     2
64                          www.worldfootball.net     1
65                                www.latimes.com     1
66                                  www.marca.com     1
67                               www.farfetch.com     1
68                            www.fcsantaclaus.fi     1
69                                  www.yahoo.com     1
70                     www.fifatrainingcentre.com     1
71         

OK - enough for URLs for now, let's have a look at users.

We will first look at users ranked by how often they've posted in our collection of reddit posts

In [115]:
user_overview_df.sort_values('total_wc_posts', ascending=False)[:100]

user_id             user_name     user_created_at  user_comments_total  \
2211     zsdhx                dotuan 2016-07-25 04:40:27                19096   
2205  ihocpit6            Meladroite 2022-01-12 00:11:53                 1683   
15    7h7t33at           MatchCaster 2020-07-31 17:01:38                 3447   
2312    14zl0j            PSGAcademy 2017-02-01 20:51:50                68184   
2233     fujeo        MatchThreadder 2014-03-25 16:08:35               175658   
984     49ocws            DrArcadeTV 2017-09-07 21:16:17                 1874   
2201  19pp70b8            dragon8811 2019-02-24 15:35:27                20299   
4455  f2iugf6r     JoJo-Bizarre-1997 2022-01-16 04:41:42                  319   
2227  2glav0rp           2soccer2bot 2018-10-22 23:50:16                17727   
101   gbbx93s4               newzee1 2021-11-07 22:52:56                  203   
2391  8jtposs4               suedney 2020-10-20 00:23:11                18750   
2416  s4xqci40           Driver_1106 2022-09-02 13:39:16                 5480   
2285  61dni8jk              rubes___ 2020-03-26 12:44:41                  264   
2308   o5vs066               Mcool18 2019-05-05 08:58:32                33095   
2293     ouhna             Blodgharm 2015-07-16 19:22:15                 1090   
2246     o1n0f       TheMonkeyPrince 2015-06-11 16:54:26                  260   
77       zisqe                 keaxy 2016-07-15 07:47:55                  209   
2292     6l4z3         AutoModerator 2012-01-05 05:24:28                  184   
2427     5wmfe          MrCrashdummy 2011-09-22 14:15:23                  676   
1087     n067h                Babbna 2015-04-19 16:45:27                   83   
2345     f1mjm           MyMoonMyMan 2014-01-30 07:43:21                 2137   
2263  6wg51d7m              alexinup 2020-06-12 12:57:29                21958   
2283     bbb8j       TheDeathOfMusic 2013-04-14 15:19:19                  156   
2255  450o8feo             moneybooy 2019-07-12 14:24:24                 1557   
2392  jzpa9z8n      RaguAlCassanello 2022-02-23 13:46:46                  196   
480   eap7giw6             cometweeb 2021-09-02 17:56:45                 6426   
7     qnlrxhx9         ForlanMatador 2022-07-30 01:52:01                  187   
2214  60er0cop            Oreallyman 2021-04-19 19:59:40                  710   
5557  e30gdxsu              Bolitico 2021-08-23 01:51:58                   60   
2503    11nyuk  Allthingsconsidered- 2016-09-25 23:10:21                  701   
2404  4bbxv4z8              Nordie27 2019-08-06 14:52:31                 1000   
2212    167j4c    OleoleCholoSimeone 2017-03-15 13:32:32                 1355   
483       None        WhyIsDobbyDead                 NaT                 3737   
226   tr9gej16  Beautiful-songlyrics 2022-10-28 10:39:01                   73   
2316     pf602             Golovking 2015-08-09 15:07:42                12231   
14    atx7px3h       KazamaShintaro7 2021-09-07 17:53:22                  117   
2695     bvkrs                 Cerxa 2013-05-31 14:12:47                  104   
4353  c5gkwbts            NoPlan4789 2021-05-15 20:01:41                   64   
482    5d0xvei            emmyxpress 2017-09-26 22:10:26                  616   
31    sfjxqo8n         AsslessBaboon 2022-09-12 18:14:03                   63   
4875  ifvdcn6d            ab24moment 2022-01-10 14:07:21                  178   
2266  lkwoixyh         companioncute 2022-04-05 15:05:37                  596   
4100  f91geh6f        DeepGreenSport 2022-09-05 15:06:31                   43   
2307  63fhxei6              twintig5 2020-04-01 17:22:34                  607   
2704      1slf                papoon 2005-11-02 05:00:00                  677   
191   21kkk27l              lala_b11 2020-05-07 18:02:57                   63   
2433    117g7x          static_reset 2016-09-06 01:47:30                  410   
2206  q66u5l92         transgal_lexi 2022-07-18 13:09:10                  190   
2259    gbzf7q            keithohara 2

Wow - so our top 5 users have all posted more than 100 times in our sample, some of them well more than 400 times. That's something like 15-20 posts on average, every single day. It will be interesting to look into what kind of posts these individuals are sharing.

In [116]:
len(user_overview_df)

7209

In [117]:
len(unique_ids)

15049

Interim summary / takeaways:

- we've collected a total of `15,049` reddit posts. 
- these posts have been shared by a total of `7209` unique users.
- some of those users have been very prolific - the max being almost 500 posts total in our entire sample
- URLs shared in the url field of our post objects are split approximately 64% (internal) to 36% external. 
- Out of external URLs, we again find a significantly large amount linking to streaming sites
- We also had a look at URLs embedded within the text of a reddit post. This gave us a further `2,773` url sharing incidences. Here, again, we see quite a few streaming-looking sites in the top domains. 

Next steps: 
- try and come up with a proportion of links to streaming sites out of urls (normal), urls (embedded) and overall. 
- have a slightly deeper look at the content (self text) posted by our most active users
- run the vader sentiment analysis on each of our reddit posts that contain self text
- produce most frequent words in positive (>0 compound) and negative (< 0 compound) posts. show as tables & as word clouds.

In [ ]:
streaming_domains = [
    'streamja.com',
    'streamin.me',
    'streamable.com',
    'streamff.com',
    'streamag.com',
    'streamingdigitally.com',
    'streamscores.link',
    'www.reddit-stream.com',
    'reddit-stream.com',
    'www.livesoccertv.com',
    'livesoccertv.com',
    'streamin.me'
]